File: bindings/java-jni/src/main/java/org/maplibre/nativejni/resource/ResourceRequest.java
package org.maplibre.nativejni.resource;

import java.util.Optional;

/** Copied network resource request passed to a runtime resource provider. */
public record ResourceRequest(
    String url,
    ResourceKind kind,
    ResourceLoadingMethod loadingMethod,
    ResourcePriority priority,
    ResourceUsage usage,
    ResourceStoragePolicy storagePolicy,
    Optional<ByteRange> range,
    Optional<Long> priorModifiedUnixMs,
    Optional<Long> priorExpiresUnixMs,
    Optional<String> priorEtag,
    byte[] priorData) {
  public ResourceRequest {
    range = range == null ? Optional.empty() : range;
    priorModifiedUnixMs = priorModifiedUnixMs == null ? Optional.empty() : priorModifiedUnixMs;
    priorExpiresUnixMs = priorExpiresUnixMs == null ? Optional.empty() : priorExpiresUnixMs;
    priorEtag = priorEtag == null ? Optional.empty() : priorEtag;
    priorData = priorData == null ? new byte[0] : priorData.clone();
  }

  @Override
  public byte[] priorData() {
    return priorData.clone();
  }

  public record ByteRange(long start, long end) {}
}

File: bindings/kotlin/src/commonMain/kotlin/org/maplibre/nativeffi/resource/ResourceRequest.kt
package org.maplibre.nativeffi.resource

/** Copied network resource request passed to a runtime resource provider. */
public class ResourceRequest(
  /** URL entering the network layer, preserving configured scheme aliases. */
  public val requestedUrl: String,
  /** URL to fetch, after tile server normalization. */
  public val resolvedUrl: String,
  public val kind: ResourceKind,
  public val loadingMethod: ResourceLoadingMethod,
  public val priority: ResourcePriority,
  public val usage: ResourceUsage,
  public val storagePolicy: ResourceStoragePolicy,
  public val range: ByteRange?,
  public val priorModifiedUnixMs: Long?,
  public val priorExpiresUnixMs: Long?,
  public val priorEtag: String?,
  priorData: ByteArray,
) {
  private val priorDataBytes: ByteArray = priorData.copyOf()

  public val priorData: ByteArray
    get() = priorDataBytes.copyOf()

  /** HTTP byte range. Values preserve native `uint64_t` bit patterns in [Long]. */
  public data class ByteRange(public val start: Long, public val end: Long)
}
